apps/examples/e2e/tests/test-canvas.spec.ts

```diff
@@ -0,0 +1,169 @@
+
+		// The shape should be selected
+		expect(await page.evaluate(() => app.selectedIds.length)).toBe(1)
+
+		// draw a box
+		await page.keyboard.press('r')
+		await page.mouse.move(10, 250)
+		await page.mouse.down()
+		expect(await page.evaluate(() => app.selectedShapes.length)).toBe(1)
+
+		// move to start drawing box (actually resizing)
+		await page.mouse.move(100, 350)
+		expect(await page.evaluate(() => app.shapesArray.length)).toBe(2)
+		expect(await page.evaluate(() => app.root.path.value)).toBe('root.select.resizing')
+
+		// finish on mouse up
+		await page.mouse.up()
+		expect(await page.evaluate(() => app.root.path.value)).toBe('root.select.idle')
+		expect(await page.evaluate(() => app.shapesArray.length)).toBe(2)
+
+		// The shape should be selected
+		expect(await page.evaluate(() => app.selectedIds.length)).toBe(1)
+	})
+
+	test('undo and redo', async ({ page }) => {
+		// buttons should be disabled when there is no history
+		expect(await page.evaluate(() => app.shapesArray.length)).toBe(0)
+		expect(await page.evaluate(() => app.selectedShapes.length)).toBe(0)
+		expect(page.getByTestId('main.undo')).toBeDisabled()
+		expect(page.getByTestId('main.redo')).toBeDisabled()
+
+		// create a shape
+		await page.keyboard.press('r')
+		await page.mouse.move(100, 100)
+		await page.mouse.down()
+		await page.mouse.up()
+
+		// We should have an undoable shape
+		expect(await page.evaluate(() => app.shapesArray.length)).toBe(1)
+		expect(await page.evaluate(() => app.selectedShapes.length)).toBe(1)
+		expect(page.getByTestId('main.undo')).not.toBeDisabled()
+		expect(page.getByTestId('main.redo')).toBeDisabled()
+
+		// Click the undo button to undo the shape
+		await page.getByTestId('main.undo').click()
+		expect(await page.evaluate(() => app.shapesArray.length)).toBe(0)
+		expect(await page.evaluate(() => app.selectedShapes.length)).toBe(0)
+		expect(page.getByTestId('main.undo')).toBeDisabled()
+		expect(page.getByTestId('main.redo')).not.toBeDisabled()
+
+		// Click the redo button to redo the shape
+		await page.getByTestId('main.redo').click()
+		expect(await page.evaluate(() => app.shapesArray.length)).toBe(1)
+		expect(await page.evaluate(() => app.selectedShapes.length)).toBe(1)
+		expect(await page.getByTestId('main.undo').isVisible()).toBe(true)
+		expect(await page.getByTestId('main.redo').isVisible()).toBe(true)
+
+		expect(await page.getByTestId('main.undo').isDisabled()).not.toBe(true)
+		expect(await page.getByTestId('main.redo').isDisabled()).toBe(true)
+	})
+
+	test('style panel + undo and redo squashing', async ({ page }) => {
+		expect(await page.getByTestId('main.undo').isVisible()).toBe(true)
+		expect(await page.getByTestId('main.redo').isVisible()).toBe(true)
+
+		await page.keyboard.press('r')
+		await page.mouse.move(100, 100)
+		await page.mouse.down()
+		await page.mouse.up()
+		expect(await page.evaluate(() => app.selectedShapes.length)).toBe(1)
+
+		const getSelectedShapeColor = async () =>
```

Comment:
Would it make more sense to actually get this from the css / html? Would make it more e2e if we check what the users sees is what we expect.

Reply:
That's a good idea!